Make bookmark view organizable

light-and-salt opened the pull request:
- [x] Allow user to re-order bookmarks using drag-and-drop
  - [x] Change `Bookmarks.dict` from `Object` to `Array`
  - [x] Clear any deprecated bookmarks from local storage. This is mostly beneficial for developers working on this project right now, who may have `Bookmarks.dict` as `Object` in their browser's local storage.
  - [x] Save the re-ordered bookmarks to local storage

![polestar](https://cloud.githubusercontent.com/assets/822034/16701857/77776c4c-4517-11e6-95ba-581df5e1d858.gif)

![voyager](https://cloud.githubusercontent.com/assets/822034/16701862/80392686-4517-11e6-9322-a17afdc683b8.gif)
- [x] When a vis is bookmarked, allow simple `textarea` annotation

![screen shot 2016-07-11 at 12 09 26 pm](https://cloud.githubusercontent.com/assets/822034/16743323/95449808-4760-11e6-83e6-1099fd11e28f.png)
- [x] Bookmark alert window
  ![bookmark alert](https://cloud.githubusercontent.com/assets/822034/16763111/40ebef6e-47dc-11e6-93be-e137bc0bad86.gif)
- [x] Tested with:
  - [x] PoleStar
  - [x] Voyager

light-and-salt committed:
Make Bookmark view organizable

Make Bookmark view organizable

Add angular-sortable-view for sortable bookmark view

Make bookmarks organizable

Clear deprecated bookmarks

Rename Bookmarks.dict to Bookmarks.list

Save Bookmark list to local storage on reorder

So that when you refresh the webpage, the order is still there

Use _.some() instead of JavaScript Array.some()

Clean up

Rename localStorage key

kanitw commented on src/services/bookmarks/bookmarks.service.js:
```diff
@@ -10,38 +10,55 @@
+     * this.dict = {} may be still in localStorage and cause problems on this.load() 
+     * This function is mostly beneficial to developers working on this repo.
+     */
+    proto.clearDeprecatedBookmarks = function() {
```
I wonder if keeping code like this is a good idea in longer term?  

Could we just use different names (e.g., `bookmarkList` ) and get rid of this entirely

light-and-salt replied:
This was mostly for my own convenience. I've cleaned it up now.

kanitw commented on src/services/bookmarks/bookmarks.service.js:
```diff
@@ -10,38 +10,55 @@
 
     proto.load = function() {
-      this.dict = localStorageService.get('bookmarks') || {};
+      this.dict = localStorageService.get('bookmarks') || [];
```
better rename this variable too.  It's not longer a `dict` but rather a `list`.

light-and-salt replied:
Done

kanitw commented on src/services/bookmarks/bookmarks.service.js:
```diff
@@ -69,15 +87,18 @@ angular.module('vlui')
 
     proto.isBookmarked = function(shorthand) {
-      return shorthand in this.dict;
+      return this.dict.some(function(bookmark) { return bookmark.shorthand === shorthand; });
```
Use `_.some` for older browser safety

light-and-salt replied:
Great tip, thanks!

light-and-salt committed:
Add annotation textarea to bookmarked views

Add annotation to bookmarked views

Sync bookmark view annotations with main view annotations

kanitw commented on src/services/bookmarks/bookmarks.service.js:
```diff
@@ -10,38 +10,53 @@
 
     proto.updateLength = function() {
-      this.length = Object.keys(this.dict).length;
+      this.length = this.list.length;
```
Why not just remove this method and property and simply use `.list.length`?

light-and-salt replied:
Done

kanitw commented on src/services/bookmarks/bookmarks.service.js:
```diff
@@ -10,38 +10,53 @@
-      this.dict = {};
+      this.list.splice(0, this.list.length);
       this.updateLength();
+      this.annotations = {};
```
Why only store the annotation part in the dict?  `isBookmarked` method can also benefit for `O(1)` check with the dictionary.

light-and-salt replied:
Done

kanitw commented:
- [x] still missing alert when people want to remove bookmark

yhoonkim commented:
I just checked it works at my local.

light-and-salt committed:
Add remove bookmark alert

light-and-salt committed:
Add placeholder to hint drop destination in bookmark view

kanitw committed:
Change placeholder style to be more in harmony with the background and make round radius to make it soft.

cc: @zeningqu

light-and-salt committed:
Maintain a dictionary of charts in memory

light-and-salt committed:
Remove bookmark updateLength() method

kanitw commented:
This is awesome.  Thanks!
